raiden/network/proxies/token.py

```diff
@@ -99,49 +101,30 @@ def approve(self, allowed_address: Address, allowance: TokenAmount) -> None:
-    ) -> str:
-        user_balance = self.balance_of(
-            address=self.client.address, block_identifier=block_identifier
-        )
-
-        # If the balance is zero, either the smart contract doesnt have a
-        # balanceOf function or the actual balance is zero
-        if user_balance == 0:
-            msg = (
-                "Approve failed. \n"
-                "Your account balance is 0 (zero), either the smart "
-                "contract is not a valid ERC20 token or you don't have funds "
-                "to use for openning a channel. "
-            )
-
-        # The approve call failed, check the user has enough balance
-        # (assuming the token smart contract may check for the maximum
-        # allowance, which is not necessarily the case)
-        elif user_balance < allowance:
-            msg = (
-                f"Approve failed. \n"
-                f"Your account balance is {user_balance}. "
-                f"The requested allowance is {allowance}. "
-                f"The smart contract may be rejecting your request due to the "
-                f"lack of balance."
-            )
-
-        # If the user has enough balance, warn the user the smart contract
-        # may not have the approve function.
-        else:
-            msg = (
-                f"Approve failed. \n"
-                f"Your account balance is {user_balance}. Nevertheless the call to"
-                f"approve failed. Please make sure the corresponding smart "
-                f"contract is a valid ERC20 token."
-            ).format(user_balance)
-
-        return msg
+                    self._raise_if_token_balance_insufficient(allowance, prefix=error_prefix)
+                    if gas_limit is None:
+                        raise RaidenRecoverableError(
+                            f"{error_prefix} Gas estimation failed for unknown reason. "
+                            f"Please make sure the contract is a valid ERC20 token."
+                        )
+                    else:
+                        raise RaidenRecoverableError(
+                            f"{error_prefix}. The reason is unknown, you have enough tokens for "
+                            f"the requested allowance and enough eth to pay the gas. There may "
+                            f"be a problem with the token contract."
+                        )
+
+    def _raise_if_token_balance_insufficient(self, required: TokenAmount, prefix: str):
+        balance = self.balance_of(self.client.address)
+
+        if balance < required:
+            msg = f"{prefix} Your balance of {balance} is below the required amount of {required}."
+            if balance == 0:
+                msg += (
+                    " Note: The balance was 0, which may also happen if the contract "
+                    "is not a valid ERC20 token (balanceOf method missing)."
+                )
+            raise InsufficientFunds(msg)
```

Comment:
This should be either a `BrokenPreconditionError` in case the balance check fail in the pre-conditions, or a `RaidenRecoverable` (recoverable because balance can change due to a race condition)

Reply:
True. I will change it to a `RaidenRecoverable` error if you agree with my reasoning above (That there are no preconditions to check for here, just race conditions).